File: PI-Frontend-main/PI-Frontend-main/src/app/formations/services/feedback.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../../environments/environment';
import { Observable } from 'rxjs';
import { Feedback, FeedbackCreatePayload, FeedbackUpdatePayload, FeedbackStats } from '../models/feedback.model';

@Injectable({ providedIn: 'root' })
export class FeedbackService {

  private readonly api = `${environment.apiUrl}/feedbacks`;

  constructor(private http: HttpClient) {}

  create(payload: FeedbackCreatePayload): Observable<Feedback> {
    return this.http.post<Feedback>(this.api, payload);
  }

  update(id: number, payload: FeedbackUpdatePayload): Observable<Feedback> {
    return this.http.put<Feedback>(`${this.api}/${id}`, payload);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.api}/${id}`);
  }

  getById(id: number): Observable<Feedback> {
    return this.http.get<Feedback>(`${this.api}/${id}`);
  }

  getByFormation(formationId: number): Observable<Feedback[]> {
    return this.http.get<Feedback[]>(`${this.api}/formation/${formationId}`);
  }

  getByCandidat(candidatId: number): Observable<Feedback[]> {
    return this.http.get<Feedback[]>(`${this.api}/candidat/${candidatId}`);
  }

  getByCandidatAndFormation(candidatId: number, formationId: number): Observable<Feedback[]> {
    return this.http.get<Feedback[]>(`${this.api}/candidat/${candidatId}/formation/${formationId}`);
  }

  getStats(formationId: number): Observable<FeedbackStats> {
    return this.http.get<FeedbackStats>(`${this.api}/formation/${formationId}/moyenne`);
  }

  getByParcours(parcoursId: number): Observable<Feedback[]> {
    return this.http.get<Feedback[]>(`${this.api}/parcours/${parcoursId}`);
  }

  getStatsForParcours(parcoursId: number): Observable<FeedbackStats> {
    return this.http.get<FeedbackStats>(`${this.api}/parcours/${parcoursId}/moyenne`);
  }

  getAll(): Observable<Feedback[]> {
    return this.http.get<Feedback[]>(this.api);
  }

  getMacrosByParcours(parcoursId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.api}/parcours/${parcoursId}/macro`);
  }

  getMacroByInscription(inscriptionId: number): Observable<any> {
    return this.http.get<any>(`${this.api}/inscription/${inscriptionId}/macro`);
  }
}

File: PI-Backend-main/PI-Backend-main/src/main/java/t/esprit/arctic/jobmatch/service/FeedbackService.java
package t.esprit.arctic.jobmatch.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import t.esprit.arctic.jobmatch.dto.FeedbackMacroDTO;
import t.esprit.arctic.jobmatch.entity.*;
import t.esprit.arctic.jobmatch.repository.*;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class FeedbackService {

    private final FeedbackRepository feedbackRepository;
    private final FeedbackMacroRepository macroRepo;
    private final InscriptionParcoursRepository inscriptionRepo;
    private final CertificatService certificatService;
    private final FeedbackAlertService alertService;
    private final NotificationService notificationService;

    public List<Feedback> getAll() {
        return feedbackRepository.findAll();
    }

    public Feedback getById(Long id) {
        return feedbackRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Feedback non trouvé avec l'id : " + id));
    }

    public Feedback create(Feedback feedback) {
        if (feedbackRepository.existsByFormationIdAndCandidatId(
                feedback.getFormation().getId(),
                feedback.getCandidat().getId())) {
            throw new RuntimeException("Vous avez déjà laissé un feedback pour cette formation");
        }
        return feedbackRepository.save(feedback);
    }

    public Feedback update(Long id, Feedback updated) {
        Feedback existing = getById(id);
        existing.setNote(updated.getNote());
        existing.setCommentaire(updated.getCommentaire());
        return feedbackRepository.save(existing);
    }

    public void delete(Long id) {
        getById(id); // vérifie l'existence
        feedbackRepository.deleteById(id);
    }


    public List<Feedback> getByFormation(Long formationId) {
        return feedbackRepository.findByFormationId(formationId);
    }

    public List<Feedback> getByCandidat(Long candidatId) {
        return feedbackRepository.findByCandidatId(candidatId);
    }

    public List<Feedback> getByCandidatAndFormation(Long candidatId, Long formationId) {
        return feedbackRepository.findByFormationIdAndCandidatId(formationId, candidatId);
    }

    public List<Feedback> getByParcours(Long parcoursId) {
        return feedbackRepository.findByParcoursId(parcoursId);
    }

    public Double getNoteMoyenne(Long formationId) {
        Double moyenne = feedbackRepository.findNoteMoyenneByFormationId(formationId);
        return moyenne != null ? Math.round(moyenne * 10.0) / 10.0 : 0.0;
    }

    @Transactional
    public void saveMacro(FeedbackMacroDTO dto) {
        InscriptionParcours inscription = inscriptionRepo.findById(dto.getInscriptionId())
                .orElseThrow(() -> new RuntimeException("Inscription non trouvée"));

        // Marquer l'exigence de feedback comme remplie (avant le check d'idempotence pour s'assurer que c'est fait)
        inscription.setEvaluationParcoursRequise(false);
        inscriptionRepo.save(inscription);

        Optional<FeedbackMacro> existingOpt = macroRepo.findByInscriptionId(dto.getInscriptionId());
        
        if (existingOpt.isPresent()) {
            FeedbackMacro macro = existingOpt.get();
            macro.setNoteGlobale(dto.getNoteGlobale());
            macro.setProgression(dto.getProgression());
            macro.setExperienceQuiz(dto.getExperienceQuiz());
            macro.setRecommandation(dto.getRecommandation());
            macro.setCommentaireLibre(dto.getCommentaireLibre());
            macroRepo.save(macro);
        } else {
            FeedbackMacro macro = FeedbackMacro.builder()
                    .inscription(inscription)
                    .candidat(inscription.getCandidat())
                    .parcours(inscription.getParcours())
                    .noteGlobale(dto.getNoteGlobale())
                    .progression(dto.getProgression())
                    .experienceQuiz(dto.getExperienceQuiz())
                    .recommandation(dto.getRecommandation())
                    .commentaireLibre(dto.getCommentaireLibre())
                    .build();
            macroRepo.save(macro);
        }

        alertService.checkLowRating(dto.getParcoursId(), "MACRO", dto.getNoteGlobale());
        
        // Générer le certificat de parcours après le feedback
        try {
            certificatService.genererPourParcours(inscription);
        } catch (Exception e) {
            System.err.println("❌ Erreur lors de la génération du certificat après feedback: " + e.getMessage());
        }
    }

    public List<FeedbackMacro> getMacroByParcours(Long parcoursId) {
        return macroRepo.findByParcoursId(parcoursId);
    }

    public Optional<FeedbackMacro> getMacroByInscriptionId(Long inscriptionId) {
        return macroRepo.findByInscriptionId(inscriptionId);
    }
}